crates/sol-macro/src/expand/contract.rs

```diff
@@ -200,6 +204,137 @@ pub(super) fn expand(cx: &ExpCtxt<'_>, contract: &ItemContract) -> Result<TokenS
         }
     });
 
+    let rpc = rpc.then(|| {
+        let contract_name = name;
+        let name = format_ident!("{contract_name}Instance");
+        let name_s = name.to_string();
+        let methods = functions.iter().map(|f| call_builder_method(f, cx));
+        let new_fn_doc = format!(
+            "Creates a new wrapper around an on-chain [`{contract_name}`](self) contract instance.\n\
+             \n\
+             See the [wrapper's documentation](`{name}`) for more details."
+        );
+        let struct_doc = format!(
+            "A [`{contract_name}`](self) instance.\n\
+             \n\
+             Contains type-safe methods for interacting with an on-chain instance of the\n\
+             [`{contract_name}`](self) contract located at a given `address`, using a given\n\
+             provider `P`.\n\
+             \n\
+             See the [module-level documentation](self) for all the available methods."
+        );
+        let (deploy_fn, deploy_method) = bytecode.is_some().then(|| {
+            let doc_str = "todo";
+            let doc = attr::mk_doc(doc_str);
+            let (params, args) = constructor.map(|c| {
+                let names1 = c.parameters.names().enumerate().map(anon_name);
+                let names2 = names1.clone();
+                let tys = c.parameters.types().map(super::ty::expand_rust_type);
+                (quote!(#(#names1: #tys),*), quote!(#(#names2,)*))
+            }).unzip();
+            (
+                quote! {
+                    #doc
+                    #[inline]
+                    pub fn deploy<P: ::alloy_contract::private::Provider>(provider: P, #params)
+                        -> ::alloy_contract::DeploymentCallBuilder<P>
+                    {
+                        #name::<P>::deploy(provider, #args)
+                    }
+                },
+                quote! {
+                    #doc
+                    #[inline]
+                    pub fn deploy(provider: P, #params)
+                        -> ::alloy_contract::DeploymentCallBuilder<P>
+                    {
+                        let data = ::alloy_sol_types::SolConstructor::abi_encode(&constructorCall { #args });
+                        ::alloy_contract::DeploymentCallBuilder::new_deployment(provider, data)
+                    }
+                },
+            )
+        }).unzip();
+        quote! {
+            #[doc = #new_fn_doc]
+            #[inline]
+            pub const fn new<P: ::alloy_contract::private::Provider>(
+                address: ::alloy_sol_types::private::Address,
```

Comment:
Does `private` work when you import from a 3rd party package?

Reply:
yes, it's `pub mod private` :smile: